File: app/src/main/java/at/shockbytes/corey/fragment/SignupFragment.java
package at.shockbytes.corey.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.firebase.ui.auth.AuthUI;
import com.firebase.ui.auth.ErrorCodes;
import com.firebase.ui.auth.IdpResponse;
import com.firebase.ui.auth.ResultCodes;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.util.Arrays;

import at.shockbytes.corey.R;
import at.shockbytes.corey.core.MainActivity;
import butterknife.ButterKnife;
import butterknife.OnClick;


public class SignupFragment extends Fragment {

    private static final int RC_SIGN_IN = 0x2341;

    public SignupFragment() {
        // Required empty public constructor
    }

    public static SignupFragment newInstance() {
        SignupFragment fragment = new SignupFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_signup, container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    @Override
    public void onStart() {
        super.onStart();

        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        if (user != null) {
            openMainActivityWithoutAnimation();
        }
    }

    @Override
    public void onDestroyView() {
        ButterKnife.unbind(this);
        super.onDestroyView();
    }

    @OnClick(R.id.fragment_signup_login)
    public void onClickSync() {

        startActivityForResult(
                AuthUI.getInstance()
                        .createSignInIntentBuilder()
                        .setProviders(Arrays.asList(new AuthUI.IdpConfig.Builder(AuthUI.EMAIL_PROVIDER).build(),
                                new AuthUI.IdpConfig.Builder(AuthUI.GOOGLE_PROVIDER).build()))
                        .setIsSmartLockEnabled(true)
                        .setLogo(R.mipmap.ic_launcher)
                        .build(),
                RC_SIGN_IN);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {

        // RC_SIGN_IN is the request code you passed into startActivityForResult(...) when starting the sign in flow.
        if (requestCode == RC_SIGN_IN) {
            IdpResponse response = IdpResponse.fromResultIntent(data);

            // Successfully signed in
            if (resultCode == ResultCodes.OK) {
                openMainActivity();
            } else {
                // Sign in failed
                if (response == null) {
                    // User pressed back button
                    showSnackbar(R.string.sign_in_cancelled);
                    return;
                }

                if (response.getErrorCode() == ErrorCodes.NO_NETWORK) {
                    showSnackbar(R.string.no_internet_connection);
                } else if (response.getErrorCode() == ErrorCodes.UNKNOWN_ERROR) {
                    showSnackbar(R.string.unknown_error);
                } else {
                    showSnackbar(R.string.unknown_sign_in_response);
                }
            }
        }
    }

    private void openMainActivity() {
        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity());
        startActivity(MainActivity.newIntent(getActivity()), options.toBundle());
        getActivity().supportFinishAfterTransition();
    }

    private void openMainActivityWithoutAnimation() {
        startActivity(MainActivity.newIntent(getActivity()));
        getActivity().supportFinishAfterTransition();
    }

    private void showSnackbar(int text) {
        Snackbar.make(getView(), text, Snackbar.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/SignUpFragment.kt
package at.shockbytes.corey.ui.fragment

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.core.app.ActivityOptionsCompat
import at.shockbytes.core.ui.fragment.BaseFragment
import at.shockbytes.corey.R
import at.shockbytes.corey.dagger.AppComponent
import at.shockbytes.corey.ui.activity.MainActivity
import com.firebase.ui.auth.AuthUI
import com.firebase.ui.auth.IdpResponse
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.fragment_signup.*

class SignUpFragment : BaseFragment<AppComponent>() {

    override val snackBarBackgroundColorRes: Int = R.color.sb_background
    override val snackBarForegroundColorRes: Int = R.color.sb_foreground

    override val layoutId = R.layout.fragment_signup

    override fun injectToGraph(appComponent: AppComponent?) {
        // Do nothing
    }

    override fun onStart() {
        super.onStart()

        // Put this call into the ViewModel
        val user = FirebaseAuth.getInstance().currentUser
        if (user != null) {
            openMainActivityWithoutAnimation()
        }
    }

    private fun onClickSignup() {

        startActivityForResult(
                AuthUI.getInstance()
                        .createSignInIntentBuilder()
                        .setAvailableProviders(
                            listOf(
                                AuthUI.IdpConfig.GoogleBuilder()
                                        .build(),
                                AuthUI.IdpConfig.EmailBuilder()
                                        .setAllowNewAccounts(true)
                                        .setRequireName(true)
                                        .build()
                            )
                        )
                        .setIsSmartLockEnabled(true)
                        .build(),
                RC_SIGN_IN
        )
    }

    override fun bindViewModel() {
    }

    override fun unbindViewModel() {
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

        // RC_SIGN_IN is the request code you passed into startActivityForResult(...) when starting the sign in flow.
        if (requestCode == RC_SIGN_IN) {
            val response = IdpResponse.fromResultIntent(data)

            // Successfully signed in
            if (resultCode == Activity.RESULT_OK) {
                openMainActivity()
            } else {
                if (response?.error != null) {
                    showSnackbar(response.error?.localizedMessage
                            ?: getString(R.string.unknown_error))
                } else {
                    // Sign in failed --> User pressed back button
                    showSnackbar(getString(R.string.sign_in_cancelled))
                }
            }
        }
    }

    private fun openMainActivity() {
        activity?.let { act ->
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(act)
            startActivity(MainActivity.newIntent(act), options.toBundle())
            act.supportFinishAfterTransition()
        }
    }

    private fun openMainActivityWithoutAnimation() {
        context?.let { ctx ->
            startActivity(MainActivity.newIntent(ctx))
            activity?.supportFinishAfterTransition()
        }
    }

    override fun setupViews() {
        fragment_signup_login.setOnClickListener {
            onClickSignup()
        }
    }

    companion object {

        private const val RC_SIGN_IN = 0x2341

        fun newInstance(): SignUpFragment {
            val fragment = SignUpFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}
